lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,256 @@
+   * compute the absolute positions (in screenshot coordinate scale) of the screenshot content
+   * and the highlighted rect around the element.
+   * @param {Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   * @param {Size} screenshotSize
+   */
+  static getScreenshotPositions(elementRectInScreenshotCoords,
+      elementPreviewSizeInScreenshotCoords, screenshotSize) {
+    const elementRectCenter = getRectCenterPoint(elementRectInScreenshotCoords);
+
+    // Try to center clipped region.
+    const screenshotLeftVisibleEdge = clamp(
+      elementRectCenter.x - elementPreviewSizeInScreenshotCoords.width / 2,
+      0, screenshotSize.width - elementPreviewSizeInScreenshotCoords.width
+    );
+    const screenshotTopVisisbleEdge = clamp(
+      elementRectCenter.y - elementPreviewSizeInScreenshotCoords.height / 2,
+      0, screenshotSize.height - elementPreviewSizeInScreenshotCoords.height
+    );
+
+    return {
+      screenshot: {
+        left: screenshotLeftVisibleEdge,
+        top: screenshotTopVisisbleEdge,
+      },
+      clip: {
+        left: elementRectInScreenshotCoords.left - screenshotLeftVisibleEdge,
+        top: elementRectInScreenshotCoords.top - screenshotTopVisisbleEdge,
+      },
+    };
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {HTMLElement} maskEl
+   * @param {{left: number, top: number}} positionClip
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   */
+  static renderClipPath(dom, maskEl, positionClip,
+      elementRectInScreenshotCoords, elementPreviewSizeInScreenshotCoords) {
+    const clipPathEl = dom.find('clipPath', maskEl);
+    const clipId = `clip-${Util.getUniqueSuffix()}`;
+    clipPathEl.id = clipId;
+    maskEl.style.clipPath = `url(#${clipId})`;
+
+    // Normalize values between 0-1.
+    const top = positionClip.top / elementPreviewSizeInScreenshotCoords.height;
+    const bottom =
+      top + elementRectInScreenshotCoords.height / elementPreviewSizeInScreenshotCoords.height;
+    const left = positionClip.left / elementPreviewSizeInScreenshotCoords.width;
+    const right =
+      left + elementRectInScreenshotCoords.width / elementPreviewSizeInScreenshotCoords.width;
+
+    const polygonsPoints = [
+      `0,0             1,0            1,${top}          0,${top}`,
+      `0,${bottom}     1,${bottom}    1,1               0,1`,
+      `0,${top}        ${left},${top} ${left},${bottom} 0,${bottom}`,
+      `${right},${top} 1,${top}       1,${bottom}       ${right},${bottom}`,
+    ];
+    for (const points of polygonsPoints) {
+      clipPathEl.append(dom.createElementNS(
+        'http://www.w3.org/2000/svg', 'polygon', undefined, {points}));
+    }
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  static createBackgroundImageStyle(dom, fullPageScreenshot) {
```

Comment:
since this needs to be called externally, would be good to have a jsdoc description for it as well